### PR TITLE
FQDN proxy keep dnscache sync

### PR DESCRIPTION
**Pull request by tgraf**
[Submiting on behalf of @eloycoto]

Two commits:

* The first one added a new functions in endpoint.
* The second one fixes the issue with FQDN restart.

Addressed remaining comments
Supersedes #6974

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6981)

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

File: pkg/endpoint/endpoint.go [hunk 1]
```diff
@@ -2295,3 +2304,41 @@ func (e *Endpoint) PinDatapathMap() error {
 
 	return err
 }
+
+func (e *Endpoint) syncEndpointHeaderFile(owner Owner) {
+	owner.GetCompilationLock().RLock()
+	defer owner.GetCompilationLock().RUnlock()
+
+	defer owner.GetCompilationLock().RUnlock()
```
Comment (ianvernon) [inline pkg/endpoint/endpoint.go]: remove this line

File: pkg/endpoint/endpoint.go [hunk 2]
```diff
@@ -2295,3 +2304,41 @@ func (e *Endpoint) PinDatapathMap() error {
 
 	return err
 }
+
+func (e *Endpoint) syncEndpointHeaderFile(owner Owner) {
+	owner.GetCompilationLock().RLock()
```
Comment (ianvernon) [inline pkg/endpoint/endpoint.go]: Shouldn't we hold the Endpoint `BuildMutex` too?

**Review by ianvernon — CHANGES_REQUESTED** [review]

**Review by raybejjani — APPROVED** [review]
We might have a race between builds that hold the compile lock and DNS updates (this would be new because we now have 2 async writers of this file but before it was always serialised via the queue). Is it possible for this to happen:
- a build is queued
- a dns update happens, this writes do disk
- the build uses an old serialisation of the endpoint and overwrites the newer DNS data

I think the answer is no, and all other order-of-operations seem safe.

File: pkg/endpoint/endpoint.go (hunk 2 above)
Comment (tgraf) [reply pkg/endpoint/endpoint.go]: Good point, I meant to lock the BuildMutex, not the CompilationLock

**Comment by tgraf** [comment]
@raybejjani The following could happen:

1. build is queued
2. DNS update happens, trigger is invoked
3. trigger is run, uses latest endpoint data
4. build is executed, uses latest endpoint data as endpoint pointer is passed in

So yes, the race can happen in terms of a DNS update preempting the build execution but I don't thikn we have stale data anywhere as we pass around the endpoint pointer. Does this match your understanding?

**Comment by raybejjani** [comment]
@tgraf yes, that is exactly what I think will happen. Should be totally fine :)

**Review by ianvernon — APPROVED** [review]

**Comment by raybejjani** [comment]
test-me-please

File: pkg/endpoint/endpoint.go [hunk 3]
```diff
@@ -2295,3 +2304,40 @@ func (e *Endpoint) PinDatapathMap() error {
...
+	}
+	defer e.Unlock()
+
+	e.writeHeaderfile(e.StateDirectoryPath())
+}
+
+// SyncEndpointHeaderFile it bumps the current DNS History information for the
+// endpoint in the lxc_config.h file.
+func (e *Endpoint) SyncEndpointHeaderFile() error {
+	e.UnconditionalLock()
```
Comment (joestringer) [inline pkg/endpoint/endpoint.go]: Can we unconditionally lock here? Sure we looked up the endpoint but could it also disappear?

File: pkg/endpoint/bpf.go [hunk 4]
```diff
@@ -127,7 +127,7 @@ func WriteIPCachePrefixes(fw *bufio.Writer, getBPFData getBPFDataCallback) {
 	}
 }
 
-func (e *Endpoint) writeHeaderfile(prefix string, owner Owner) error {
+func (e *Endpoint) writeHeaderfile(prefix string) error {
```
Comment (joestringer) [inline pkg/endpoint/bpf.go]: nit: Removing owner is not part of the fix. I think I'll need it again soon for the templating stuff too, but I can always rebase and reinstate it.

**Review by joestringer — CHANGES_REQUESTED** [review]
Looks good, except the unconditional lock.

**Comment by tgraf** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]

**Comment by tgraf** [comment]
`runtime.RuntimeFQDNPolicies Roundrobin DNS` failed

EDIT: Wrong, I pressed back one too many times and ended up putting the test failure of #6981 into this PR.

**Comment by tgraf** [comment]
test-me-please

**Comment by ianvernon** [comment]
Build failure: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/9694/testReport/junit/(root)/runtime/RuntimeFQDNPolicies_Roundrobin_DNS/

Round robin DNS test failed again :/

**Comment by ianvernon** [comment]
test-me-please

**Pull request merged**